common/flatpak-installation.c

```diff
@@ -991,6 +991,75 @@ _ostree_collection_ref_free0 (OstreeCollectionRef *ref)
   ostree_collection_ref_free (ref);
 }
 
+static gboolean
+flatpak_installation_check_installed_ref_missing_related_ref (FlatpakInstallation *self,
+                                                              const gchar         *remote_name,
+                                                              gchar               *full_ref,
+                                                              GCancellable        *cancellable)
+{
+  g_autoptr (GPtrArray) installed_related_refs = NULL;
+  g_autoptr (GPtrArray) remote_related_refs = NULL;
+  g_autoptr (GError) local_error = NULL;
+  guint j, k;
+
+  installed_related_refs = flatpak_installation_list_installed_related_refs_sync (self,
+                                                                                  remote_name,
+                                                                                  full_ref,
+                                                                                  cancellable,
+                                                                                  &local_error);
+  if (local_error != NULL)
+    {
+      g_warning ("Unable to get installed related refs for %s: %s", full_ref, local_error->message);
+      return FALSE;
+    }
+
+  remote_related_refs = flatpak_installation_list_remote_related_refs_sync (self,
+                                                                            remote_name,
+                                                                            full_ref,
+                                                                            cancellable,
+                                                                            &local_error);
+  if (local_error != NULL)
+    {
+      g_warning ("Unable to get remote related refs for %s: %s", full_ref, local_error->message);
+      return FALSE;
+    }
+
+  if (remote_related_refs == NULL)
+    return FALSE;
+
+  for (j = 0; j < remote_related_refs->len; j++)
+    {
+      FlatpakRelatedRef *remote_related_ref = g_ptr_array_index (remote_related_refs, j);
+      g_autofree gchar *remote_related_full_ref = flatpak_ref_format_ref (FLATPAK_REF (remote_related_ref));
+      gboolean should_download = flatpak_related_ref_should_download (remote_related_ref);
+      gboolean related_ref_found = FALSE;
+
+      if (should_download && installed_related_refs != NULL)
+        {
+          for (k = 0; k < installed_related_refs->len; k++)
+            {
+              FlatpakRelatedRef *installed_related_ref = g_ptr_array_index (installed_related_refs, k);
+              g_autofree gchar *installed_related_full_ref = flatpak_ref_format_ref (FLATPAK_REF (installed_related_ref));
+
+              if (g_strcmp0 (installed_related_full_ref, remote_related_full_ref) == 0)
```

Comment:
Could this be made to compare the already-decomposed fields of `installed_related_ref` and `remote_related_ref`, rather than turning both back into strings and comparing those?

Reply:
My idea was to get as strict checking as possible for the refs. If upstream maintainers agree that comparing decomposed refs is good enough checking, I don't have much problem with it.

Reply:
Why would comparing the parsed form be any different to the string form?

Reply:
> Could this be made to compare the already-decomposed fields of `installed_related_ref` and `remote_related_ref`, rather than turning both back into strings and comparing those?

I realized that the fields are not already string-decomposed at the first place. Both are `FlatpakRelatedRefs` pointers.